### PR TITLE
Add support for azure keyvault

### PR DESCRIPTION
**Pull request by jdamata**
Im not much of a dev nor have any experience with js/nodejs. Please tear this PR to shreds :D

### CODE REVIEW

File: config/azure-config.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+'use strict'
+
+const { DefaultAzureCredential } = require('@azure/identity')
+// DefaultAzureCredential expects the following three environment variables:
+// - AZURE_TENANT_ID: The tenant ID in Azure Active Directory
+// - AZURE_CLIENT_ID: The application (client) ID registered in the AAD tenant
+// - AZURE_CLIENT_SECRET: The client secret for the registered application
+
+module.exports = {
+  azureKeyVault: () => {
```
[the remark this answers is not in the document]
Comment (jdamata) [reply config/azure-config.js]: Im not sure what to name it. Suggestion?

**Comment by jdamata** [comment]
I noticed that the status of a failed externalsecrets for this backend looks a bit too verbose. I don't know how to change this though.

```
$ kubectl get externalsecret
NAME               LAST SYNC   STATUS   AGE
keyvault-example   3m47s       ERROR, Authentication failed to complete due to the following errors:

AuthenticationError: An error was returned while authenticating to Azure Active Directory (status code 400).

More details:

{
  "error": "missing_environment_variables",
  "errorDescription": "EnvironmentCredential cannot return a token because one or more of the following environment variables is missing:\n\nAZURE_TENANT_ID\nAZURE_CLIENT_ID\nAZURE_CLIENT_SECRET\nAZURE_CLIENT_CERTIFICATE_PATH\nAZURE_USERNAME\nAZURE_PASSWORD\n\nTo authenticate with a service principal AZURE_TENANT_ID, AZURE_CLIENT_ID, and either AZURE_CLIENT_SECRET or AZURE_CLIENT_CERTIFICATE_PATH must be set.  To authenticate with a user account AZURE_TENANT_ID, AZURE_USERNAME, and AZURE_PASSWORD must be set.\n"
}                  3m52s
```

File: config/azure-config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Flydiverny) [reply config/azure-config.js]: Let's leave it as is. :) not sure what I had in mind

**Comment by Flydiverny** [comment]
Ah the verbose errors could be abit annoying, but I think that's outside the scope of this PR. Think we should probably rework how those are in some way. Been thinking about if one can create events for the error message instead and just write error as a status, or similar.

**Comment by silasbw** [comment]
We just released 3.1.0 and included this change.  Thanks again for the contribution :medal_sports: :100:

**Comment by mo-saeed** [comment]
@jdamata Is there any way to restrict k8s namespace to have only access to a specific subscription?

same way as assuming_roles here https://github.com/godaddy/kubernetes-external-secrets#using-aws-access-credentials

**Comment by jdamata** [comment]
It sounds like you are mapping a k8s namespace to a keyvault in a subscription.

Currently the kubernetes-external-secrets operator takes a single azure service principal to authenticate against azure keyvaults.

At this point I don't think you can achieve what you are asking. There can only be one operator deployed and it will need SP credentials that has access to get keys from the keyvaults that you are interested in.

Some potentials ways to get what you want would be:
- Deploy a kubernetes-external-secrets per namespace that has a service principal with permissions to only get keys from specific keyvaults (Requires https://github.com/godaddy/kubernetes-external-secrets/pull/193)
- Add the ability to override the service principal deployed with the operator in each ExternalSecret resource. Then you could specify per ExternalSecret what service principal to use.

**Pull request merged**